sdk/storage/azure-storage-files-datalake/src/datalake_directory_client.cpp

```diff
@@ -220,21 +220,40 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
     protocolLayerOptions.Recursive = recursive;
     protocolLayerOptions.ContinuationToken = options.ContinuationToken;
 
-    const std::string currentPath = m_pathUrl.GetPath();
-    auto firstSlashPos = std::find(currentPath.begin(), currentPath.end(), '/');
-    const std::string fileSystemName(currentPath.begin(), firstSlashPos);
-    if (firstSlashPos != currentPath.end())
+    Azure::Core::Url fileSystemUrl;
+    if (m_clientConfiguration.FileSystemUrl.HasValue())
     {
-      ++firstSlashPos;
+      fileSystemUrl = m_clientConfiguration.FileSystemUrl.Value();
+      const std::string fileSystemPath = fileSystemUrl.GetPath();
+      const std::string currentPath = m_pathUrl.GetPath();
+      std::string directoryPath = currentPath.substr(fileSystemPath.length());
+      if (directoryPath.length() > 0 && directoryPath[0] == '/')
```

Comment:
Yes. The contract is:

* if customer provides a full URL to storage, it should be URL-encoded (`%` + 2-char hex in UTF-8 encoding).
* if customer provides a path segment (like providing directory name when creating a directory client from filesystem client, the directory name is a path segment), it's not URL-encoded. Storage SDK encodes the path segment and appends it to existing URL.

Reply:
> If the path is NOT url encoded, then what is the encoding of the path?

The path segment must be in UTF-8 encoding, no matter what locale is. That's the contract. If the customer's locale is not UTF-8, it's their responsibility to convert path segment to UTF-8 before passing to the SDK.

> Depending on the encoding, it may not be safe to look at arbitrary characters in the string for certain characters.

After the url and path segments are processed by public API, they're always URL-encoded in private variables. This simplifies our internal code. And it's safe to search for ascii characters.

Reply:
IMHO there should be no circumstance where a part of the path might be UTF-8 encoded but other parts are URL encoded. Strings should ahve a single encoding for the entire string.

As long as there's a well-defined encoding of the string types, and the defined encoding is NOT a WIndows SBCS encoding, I'm ok.

Reply:
> IMHO there should be no circumstance where a part of the path might be UTF-8 encoded but other parts are URL encoded. Strings should ahve a single encoding for the entire string.

This would never happen in our SDK. I'd like to clarify with an example,

the customer creates a directory client with url `https://account.dfs.core.windows.net/filesystem/ā`. The URL string passed to our SDK should be `https://account.dfs.core.windows.net/filesystem/%C4%81`.
Now this customer wants to get a subdirectory (`ć`) under the previous directory, he should call `directoryClient->GetSubdirectoryClinet("\xC4\x87")`. Our SDK will encode the new path segment with URL-encoding, now the url in private variable is `https://account.dfs.core.windows.net/filesystem/%C4%81/%C4%87`.

Reply:
Ah, that helps a lot. I got confused by this comment:
> if customer provides a path segment (like providing directory name when creating a directory client from filesystem client, the directory name is a path segment), it's not URL-encoded. Storage SDK encodes the path segment and appends it to existing URL.

When I read it, I missed the "Storage SDK encodes the path segment" comment.

Silly question: What encoding does the customer specify when they provide a directory name? Does the documentation make it clear that it MUST be UTF-8 and not Windows-ACP or any other Windows MBCS encoding?

Reply:
I'll document it.

https://github.com/Azure/azure-sdk-for-cpp/issues/4936